mongodb/vibe/db/mongo/collection.d

```diff
@@ -434,106 +436,7 @@ struct MongoCollection {
-
-		Note that the overload taking an associative array of field orders
-		will be removed. Since the order of fields matters, it is
-		only suitable for single-field indices.
-	*/
-	void ensureIndex(scope const(Tuple!(string, int))[] field_orders, IndexFlags flags = IndexFlags.none, Duration expire_time = 0.seconds)
-	@safe {
-		// TODO: support 2d indexes
-
-		auto key = Bson.emptyObject;
-		auto indexname = appender!string();
-		bool first = true;
-		foreach (fo; field_orders) {
-			if (!first) indexname.put('_');
-			else first = false;
-			indexname.put(fo[0]);
-			indexname.put('_');
-			indexname.put(to!string(fo[1]));
-			key[fo[0]] = Bson(fo[1]);
-		}
-
-		Bson[string] doc;
-		doc["v"] = 1;
-		doc["key"] = key;
-		doc["ns"] = m_fullPath;
-		doc["name"] = indexname.data;
-		if (flags & IndexFlags.unique) doc["unique"] = true;
-		if (flags & IndexFlags.dropDuplicates) doc["dropDups"] = true;
-		if (flags & IndexFlags.background) doc["background"] = true;
-		if (flags & IndexFlags.sparse) doc["sparse"] = true;
-		if (flags & IndexFlags.expireAfterSeconds) doc["expireAfterSeconds"] = expire_time.total!"seconds";
-		database["system.indexes"].insert(doc);
-	}
-
-	/// ditto
-	deprecated("Use the overload taking an array of field_orders instead.")
-	void ensureIndex(int[string] field_orders, IndexFlags flags = IndexFlags.none, ulong expireAfterSeconds = 0)
-	@safe {
-		Tuple!(string, int)[] orders;
-		foreach (k, v; field_orders)
-			orders ~= tuple(k, v);
-		ensureIndex(orders, flags, expireAfterSeconds.seconds);
-	}
-
-	/**
-		Drops or removes the specified index from the collection.
-	*/
-	void dropIndex(string name)
-	@safe {
-		static struct CMD {
-			string dropIndexes;
-			string index;
-		}
-
-		CMD cmd;
-		cmd.dropIndexes = m_name;
-		cmd.index = name;
-		auto reply = database.runCommand(cmd);
-		enforce(reply["ok"].get!double == 1, "dropIndex command failed: "~reply["errmsg"].opt!string);
-	}
-
-	/**
-		Creates indexes on the collection.
-	*/
-	void createIndex(T)(T query) 
-	@safe {
-		static struct Indexes {
-			T key;
-		}
-
-		static struct CMD {
-			string createIndexes;
-			Indexes indexes;
-		}
-
-		CMD cmd;
-		cmd.createIndexes = m_name;
-		cmd.indexes.key = query;
-		auto reply = database.runCommand(cmd);
-		enforce(reply["ok"].get!double == 1, "createIndex command failed: "~reply["errmsg"].opt!string);
-	}
-
-	/**
-		Returns an array that holds a list of documents that identify and describe the existing indexes on the collection. 
-	*/
-	MongoCursor!R getIndexes(T = Bson, R = Bson)() 
-	@safe {
-		static struct CMD {
-			string listIndexes;
-		}
-
-		CMD cmd;
-		cmd.listIndexes = m_name;
-
-		auto reply = database.runCommand(cmd);
-		enforce(reply["ok"].get!double == 1, "getIndexes command failed: "~reply["errmsg"].opt!string);
-		return MongoCursor!R(m_client, reply["cursor"]["ns"].get!string, reply["cursor"]["id"].get!long, reply["cursor"]["firstBatch"].get!(Bson[]));
-	}
+	mixin MongoCollectionIndexStandardAPIImpl;
```

Comment:
Thanks for the review! I think using mixin templates here is really the D way to go. When I plan to implement the other APIs (see #2200) it would massively fill this collection.d file

The inline documentation & type definitions alone will probably make up over half of the file then. Having them in dedicated modules improves code maintainability imo.

The MongoDB driver spec also has separate documents for all the different features a driver may implement, having corresponding D modules for this is another plus.

Having separate implementation files is okay, but for example you would either need to omit the documentation for the methods or duplicate the documentation in the documentation file. I think an attractive pro argument of using the mixin templates is that you have both code and documentation in one place without duplication, which makes development a little more convenient. This is kind of D's equivalent to C#'s partial classes.

So I think we should be using D's great features here and if there is a problem with tooling/documentation rather fix that to support D better.

Reply:
There is also the alternative of using UFCS and freestanding functions in a separate module. As far as I can see, that would solve both issues.

Reply:
hm I'm not a big fan of UFCS for this kind of functionality because it often breaks discoverability of the functions (auto complete, documentation)

I tested now how the documentation output looks in ddox: for the mixin template it looks just like a class with all the member functions, which I think is pretty good and works just fine.

How about we just link to all mixin'd templates at the top of the class definition?

Reply:
I would usually expect a mixin to have worse issues with auto-completion than a UFCS function, but I have no idea how this is distributed across real-world implementations. It's good to see that the mixed-in members appear correctly in the documentation, but I still don't really like this stylistically and would have to think more about it before coming to a final opinion.

So then let's just keep the methods as normal members for now, so that the non-controversial parts can be merged.

Reply:
moved all the members into the class now so it's uncontroversial